prow/interrupts/interrupts.go

```diff
@@ -0,0 +1,204 @@
+var single *manager
+
+// WaitForGracefulShutdown waits until all registered servers and workers
+// have had time to gracefully shut down, or times out. This function is
+// blocking.
+func (m *manager) WaitForGracefulShutdown() {
+	finished := make(chan struct{})
+	go func() {
+		defer close(finished)
+		m.wg.Wait()
+	}()
+	select {
+	case <-finished:
+		logrus.Info("All workers gracefully terminated, exiting.")
+	case <-time.After(1 * time.Minute):
+		logrus.Warn("Timed out waiting for workers to gracefully terminate, exiting.")
+	}
+}
+
+// Context returns a context that stays is cancelled when an interrupt hits.
+// Using this context is a weak guarantee that your work will finish before
+// process exit as callers cannot signal that they are finished. Prefer to use
+// Run().
+func (m *manager) Context() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		m.wg.Add(1)
+		defer m.wg.Done()
+
+		m.c.L.Lock()
+		m.c.Wait()
+		cancel()
+		m.c.L.Unlock()
+	}()
+
+	return ctx
+}
+
+// Tick will do work until an interrupt is received, then signal the
+// worker. This function is not blocking. Callers are expected to exit
+// only after WaitForGracefulShutdown returns to ensure all workers have
+// had time to shut down. This is preferrable to getting the raw Context
+// as we can ensure that the work is finished before releasing our share
+// of the wait group on shutdown.
+func (m *manager) Run(work func(ctx context.Context) error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		m.wg.Add(1)
+		defer m.wg.Done()
+		if err := work(ctx); err != nil {
+			logrus.WithError(err).Error("Failed to shut down worker after interrupt.")
+		}
+	}()
+
+	m.c.L.Lock()
+	m.c.Wait()
+	cancel()
+	m.c.L.Unlock()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServe(server *http.Server, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServe()).Info("Server exited.")
+	}()
+}
+
+// ListenAndServe runs the HTTP server and handles shutting it down
+// gracefully on interrupts. This function is not blocking. Callers
+// are expected to exit only after WaitForGracefulShutdown returns to
+// ensure all servers have had time to shut down.
+func (m *manager) ListenAndServeTLS(server *http.Server, certFile, keyFile string, gracePeriod time.Duration) {
+	go m.shutdown(server, gracePeriod)
+	go func() {
+		logrus.WithError(server.ListenAndServeTLS(certFile, keyFile)).Info("Server exited.")
+	}()
+}
+
+// shutdown will shut down the server when an interrupt is received
+func (m *manager) shutdown(server *http.Server, gracePeriod time.Duration) {
+	m.wg.Add(1)
+
+	m.c.L.Lock()
+	m.c.Wait()
```

Comment:
Technically there could be a broadcast to the condition variable before we are even able to lock it and start waiting. To prevent this we'd need to add a bool to `manager` that indicates if we've seen an interrupt or not and check it after locking, but before calling `Wait()` here. This also applies to multiple places.